Use TransitionInfo in more interfaces 3

wddgit opened the pull request:
#### PR description:

Use the TransitionInfo classes in more interfaces. In particular, it adds interface changes to streamTransitionAsync.h that were recently made in globalTransitionAsync.h so the two will remain similar. It also changes other interfaces from EventProcessor down the stack to the Worker classes and into the ProductResolvers.

#### PR validation:

Ran existing unit tests. This should not change any behavior or any results. It just makes the code easier to read and understand.

(This is the third try at this. I incorporated the changes we discussed this morning on zoom. In particular, you might want to look at ProductResolvers.cc. Also, I renamed one of the setupOnDemandSystem in WorkerManager. The rest is almost identical to the previous version of the PRs. I removed the percolation code in prefetchAsync_ and that simplified things. I also rebased and squashed the whole thing into one commit.)

wddgit committed:
Use TransitionInfo in more interfaces

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

ERROR: Build errors found during clang-tidy run.
```
FWCore/TestProcessor/src/TestProcessor.cc:46:10: error: 'FWCore/Concurrency/interface/ThreadsController.h' file not found [clang-diagnostic-error]
#include "FWCore/Concurrency/interface/ThreadsController.h"
         ^
Suppressed 1321 warnings (1320 in non-user code, 1 with check filters).
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:128: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

wddgit commented:
code-checks

Try again, I think there was a change not yet in the IB used as a base for the first attempt that was needed

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

ERROR: Build errors found during clang-tidy run.
```
FWCore/TestProcessor/src/TestProcessor.cc:46:10: error: 'FWCore/Concurrency/interface/ThreadsController.h' file not found [clang-diagnostic-error]
#include "FWCore/Concurrency/interface/ThreadsController.h"
         ^
Suppressed 1321 warnings (1320 in non-user code, 1 with check filters).
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:128: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

wddgit commented:
code-checks

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31442/18329

- This PR adds an extra 12KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Framework/interface/Schedule.h  modified in PR(s): #29553
  - File  FWCore/Framework/interface/UnscheduledCallProducer.h  modified in PR(s): #29553
  - File  FWCore/Framework/interface/WorkerManager.h  modified in PR(s): #29553
  - File  FWCore/Framework/src/EventProcessor.cc  modified in PR(s): #29553
  - File  FWCore/Framework/src/GlobalSchedule.h  modified in PR(s): #29553
  - File  FWCore/Framework/src/Path.cc  modified in PR(s): #29553
  - File  FWCore/Framework/src/ProductResolvers.cc  modified in PR(s): #29553
  - File  FWCore/Framework/src/Schedule.cc  modified in PR(s): #29553
  - File  FWCore/Framework/src/StreamSchedule.cc  modified in PR(s): #29553
  - File  FWCore/Framework/src/StreamSchedule.h  modified in PR(s): #29553
  - File  FWCore/Framework/src/SubProcess.cc  modified in PR(s): #29553
  - File  FWCore/Framework/src/Worker.cc  modified in PR(s): #29553
  - File  FWCore/Framework/src/Worker.h  modified in PR(s): #29553
  - File  FWCore/Framework/src/WorkerManager.cc  modified in PR(s): #29553

cmsbuild commented:
A new Pull Request was created by  @wddgit (W. David Dagenhart) for master.

It involves the following packages:

FWCore/Framework
FWCore/TestProcessor
Mixing/Base


@smuzaffar, @civanch, @Dr15Jones, @makortel, @mdhildreth, @cmsbuild can you please review it and eventually sign? Thanks.
@makortel, @fabiocos this is something you requested to watch as well.
@silviodonato, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

wddgit commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: a45616165ff7f60ea71004aad44af276b30df03c
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f80f0f/9291/summary.html
CMSSW: CMSSW_11_2_X_2020-09-12-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f80f0f/9291/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2620306
 - DQMHistoTests: Total failures: 5
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 2620278
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.004 KiB( 34 files compared)
 - DQMHistoSizes: changed ( 10224.0 ): -0.004 KiB       MessageLogger/Warnings
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

makortel commented:
> This is the third try at this. 

Just to record for any future readers, the earlier two attempts are #30969 and #31136.

makortel commented:
Looks good to me. @Dr15Jones, do you have any comments?

wddgit commented:
There is one additional comment I should make on this PR. There is still one more TransitionInfo PR coming after this one that extends this into the code below the Worker class in the stack. That PR is already written. I just need to rebase it after this one gets merged.

Dr15Jones commented on FWCore/Framework/src/Worker.h:
```diff
@@ -645,13 +618,19 @@ namespace edm {
-        return iWorker->implDo(ep, es, mcc);
+        return iWorker->implDo(info.principal(), info.eventSetupImpl(), mcc);
+      }
+      static void esPrefetch(Worker* worker,
```
Please add `Async` to the name to make it clear this is not a synchronous call.

wddgit replied:
Thanks. That is better. I am changing this now and also all the similar places referenced by the the other comments. I will push the commit soon.

Dr15Jones commented on FWCore/Framework/src/Worker.h:
```diff
@@ -666,13 +645,19 @@ namespace edm {
-        return iWorker->implDoBegin(ep, es, mcc);
+        return iWorker->implDoBegin(info.principal(), info.eventSetupImpl(), mcc);
+      }
+      static void esPrefetch(Worker* worker,
```
Please add Async to the name to make it clear this is not a synchronous call.

Dr15Jones commented on FWCore/Framework/src/Worker.h:
```diff
@@ -685,13 +670,19 @@ namespace edm {
-        return iWorker->implDoStreamBegin(id, ep, es, mcc);
+        return iWorker->implDoStreamBegin(id, info.principal(), info.eventSetupImpl(), mcc);
+      }
+      static void esPrefetch(Worker* worker,
```
Please add Async to the name to make it clear this is not a synchronous call.

Dr15Jones commented on FWCore/Framework/src/Worker.h:
```diff
@@ -704,13 +695,19 @@ namespace edm {
-        return iWorker->implDoEnd(ep, es, mcc);
+        return iWorker->implDoEnd(info.principal(), info.eventSetupImpl(), mcc);
+      }
+      static void esPrefetch(Worker* worker,
```
Please add Async to the name to make it clear this is not a synchronous call.

Dr15Jones commented on FWCore/Framework/src/Worker.h:
```diff
@@ -836,17 +864,18 @@ namespace edm {
+        return iWorker->implDoAccessInputProcessBlock(info.principal(), mcc);
       }
+      static constexpr void esPrefetch(
+          Worker*, WaitingTask*, ServiceToken const&, ProcessBlockTransitionInfo const&, Transition) {}
```
You need to check that the calls into here are not expecting this module to run the task if the task is at ref count 0.

wddgit replied:
I think this is OK. This function is only called in one place. The ref count is incremented before the function is called and then decremented and checked after the function is called. Does that cover the case you are worried about?

```
    //Need to be sure the ref count isn't set to 0 immediately
    iTask->increment_ref_count();

    workerhelper::CallImpl<T>::esPrefetchAsync(this, iTask, token, transitionInfo, iTransition);
    edPrefetchAsync(iTask, token, principal);

    if (principal.branchType() == InEvent) {
      preActionBeforeRunEventAsync(iTask, moduleCallingContext_, principal);
    }

    if (0 == iTask->decrement_ref_count()) {
      //if everything finishes before we leave this routine, we need to launch the task
      tbb::task::spawn(*iTask);
    }
```

Dr15Jones replied:
agreed

Dr15Jones commented on FWCore/Framework/src/WorkerManager.cc:
```diff
@@ -126,14 +126,17 @@ namespace edm {
-    unscheduled_.setEventSetup(es);
     if (&ep != lastSetupEventPrincipal_) {
       UnscheduledConfigurator config(allWorkers_.begin(), allWorkers_.end(), &(unscheduled_.auxiliary()));
       ep.setupUnscheduled(config);
```
This stuff is still part of the 'on demand' system so the split of function names isn't really accurate.

wddgit replied:
You are partially correct because this is important for the 'on demand' system, but it is also important for other things and would be needed even of all producers were scheduled.

resetAll affects all types of resolvers, not just the unscheduled resolvers.

Four types of ProductResolvers have data members which are set by the call to setupUnscheduled. For PuttableProductResolvers, this sets worker_, which is used to notify other modules when a scheduled module is finished prefetching. This would be needed even if all producers were scheduled. For InputProductResolvers, it is setting up signals related to delayed read. I agree that in the case of UnscheduledProductResolvers it is part of the "on demand" system. I don't know how SwitchBaseProductResolver works, so I am not sure whether that is related to "on demand" or not.

wddgit committed:
Change function name to esPrefetchAsync

wddgit commented:
code-checks

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31442/18450

- This PR adds an extra 24KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Framework/interface/Schedule.h  modified in PR(s): #29553
  - File  FWCore/Framework/interface/UnscheduledCallProducer.h  modified in PR(s): #29553
  - File  FWCore/Framework/interface/WorkerManager.h  modified in PR(s): #29553
  - File  FWCore/Framework/src/EventProcessor.cc  modified in PR(s): #29553
  - File  FWCore/Framework/src/GlobalSchedule.h  modified in PR(s): #29553
  - File  FWCore/Framework/src/Path.cc  modified in PR(s): #29553
  - File  FWCore/Framework/src/ProductResolvers.cc  modified in PR(s): #29553
  - File  FWCore/Framework/src/Schedule.cc  modified in PR(s): #29553
  - File  FWCore/Framework/src/StreamSchedule.cc  modified in PR(s): #29553
  - File  FWCore/Framework/src/StreamSchedule.h  modified in PR(s): #29553
  - File  FWCore/Framework/src/SubProcess.cc  modified in PR(s): #29553
  - File  FWCore/Framework/src/Worker.cc  modified in PR(s): #29553
  - File  FWCore/Framework/src/Worker.h  modified in PR(s): #29553
  - File  FWCore/Framework/src/WorkerManager.cc  modified in PR(s): #29553

cmsbuild commented:
Pull request #31442 was updated. @smuzaffar, @civanch, @Dr15Jones, @makortel, @mdhildreth, @cmsbuild can you please check and sign again.

wddgit commented:
please test

I think I addressed all the comments from Chris. Please let me know exactly what you want me to do if you would like me to change something more, especially where I responded to his comments instead of making a change.

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 1e0348039250b461f3deda90469b113e09ad67bb
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f80f0f/9416/summary.html
CMSSW: CMSSW_11_2_X_2020-09-17-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-f80f0f/9416/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2540471
 - DQMHistoTests: Total failures: 7
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2540442
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

wddgit commented:
I think this one is ready to be approved as is... Let me know if there is something more that needs to be done. I can rename the function setupResolvers (please give me a suggestion) or recombine those functions into one using templates or modify this part of the code in some other way. One could address naming issues in that part of the code in general, although that might also be better left to a separate PR... I looked at the concurrency issue Chris mentioned and I think the function that calls the referenced code already takes care of the issue he was concerned about.

makortel commented:
+1

wddgit commented:
The simulation related changes are all in one file, Mixing/Base/src/SecondaryEventProvider.cc. And they are just reacting to the changes in the Framework interfaces. This is the only place outside FWCore/Framework that is affected.

civanch commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

silviodonato commented:
+1
